osu.Game/Database/RealmAccess.cs

```diff
@@ -388,26 +388,46 @@ public void Write(Action<Realm> action)
             }
         }
 
+        private readonly CountdownEvent pendingAsyncWrites = new CountdownEvent(0);
+
         /// <summary>
         /// Write changes to realm asynchronously, guaranteeing order of execution.
         /// </summary>
         /// <param name="action">The work to run.</param>
         public Task WriteAsync(Action<Realm> action)
         {
-            // Regardless of calling Realm.GetInstance or Realm.GetInstanceAsync, there is a blocking overhead on retrieval.
-            // Adding a forced Task.Run resolves this.
+            if (isDisposed)
+                throw new ObjectDisposedException(nameof(RealmAccess));
+
+            // Required to ensure the write is tracked and accounted for before disposal.
+            // Can potentially be avoided if we have a need to do so in the future.
+            if (!ThreadSafety.IsUpdateThread)
+                throw new InvalidOperationException(@$"{nameof(WriteAsync)} must be called from the update thread.");
 
-            return Task.Run(async () =>
+            lock (realmLock)
             {
-                total_writes_async.Value++;
+                // CountdownEvent will fail if already at zero.
+                if (!pendingAsyncWrites.TryAddCount())
+                    pendingAsyncWrites.Reset(1);
```

Comment:
`.TryAddCount(int)` is [documented as not actually being thread-safe](https://docs.microsoft.com/en-us/dotnet/api/system.threading.countdownevent.tryaddcount?view=net-6.0#system-threading-countdownevent-tryaddcount(system-int32)), which is a problem given that the `.Dispose()` usage site is using it without further locking. Any reason this isn't just [a simple `.AddCount()`?](https://docs.microsoft.com/en-us/dotnet/api/system.threading.countdownevent.addcount?view=net-6.0#system-threading-countdownevent-addcount) That method doesn't check if the count is already zero, at least not as far as I can tell from xmldoc.

Reply:
Literally no reason. I just see a `Try` method and assume it is the better of the two, based on prior knowledge. Feel free to change it if it works and passes tests, seems good.

Reply:
`AddCount` does seem to fail:

![JetBrains Rider 2022-06-28 at 08 11 20](https://user-images.githubusercontent.com/191335/176128463-bd13c6eb-e6d5-40e0-8ed2-535122575aa3.png)